### ISSUE TITLE
Add TypeScript typings to the react package 

### ISSUE THREAD

**Issue by AustinFelipe** [uNmAnNeR/imaskjs#329]
I'll be nice to have TypeScript typings for react package as well.

**Comment by timbru31** [comment]
Same goes for Angular

**Comment by mattcosta7** [comment]
A pass at some i'm starting to use - might be a good starting point

```
declare module "react-imask" {
  import Imask from "imask";

  /**
   * The props accepted by react-imask, based
   * on the implementation of imask, with  some additions
   */
  export type IMaskInputProps = Partial<
    Pick<
      Imask.AnyMaskedOptions,
      | "mask"
      | "value"
      | "prepare"
      | "validate"
      | "commit"
      | "overwrite"
      | "placeholderChar"
      | "lazy"
      | "definitions"
      | "blocks"
      | "pattern"
      | "format"
      | "parse"
      | "autofix"
      | "radix"
      | "thousandsSeparator"
      | "mapToRadix"
      | "scale"
      | "signed"
      | "normalizeZeros"
      | "padFractionalZeros"
      | "min"
      | "max"
      | "dispatch"
    >
  > & {
    unmask?: boolean;
    onAccept?: <T>(...args: T[]) => void;
    onComplete?: <T>(...args: T[]) => void;
  };

  /**
   * A function that decorates a react component
   * with 'imask' props
   * @param Component Any React Component
   */
  export function IMaskMixin<T, D>(
    Component: React.ComponentType<{ inputRef: React.Ref<D> } & T>
  ): React.ComponentType<T & IMaskInputProps>;

  /**
   * A basic IMask React Input
   */
  export const IMaskInput: React.ComponentType<IMaskInputProps>;
}

```

**Comment by hakimio** [comment]
`imask` has typings which you can just import directly from `imask` package:
```ts
import IMask from 'imask';

const dateMaskOptions: IMask.MaskedDateOptions = {};
```

**Comment by OmarOmeiri** [comment]
I'm really new to Typescript, so be gentle. Here is what I've come up with.

```
declare module 'react-imask' {
  import Imask from 'imask';

  /**
   * Missing some types here, It would be great if someone
   * stepped in.
   */
  type masked<T> =
    T extends DateConstructor
    ? Imask.MaskedDate
    : T extends NumberConstructor
    ? Imask.MaskedNumber
    : T extends RegExpConstructor
    ? Imask.MaskedRegExp
    : T extends Function
    ? Imask.MaskedFunction
    : Imask.Masked<IMaskInputProps['mask']>

  interface InputMask <T extends Imask.AnyMaskedOptions> {
    el: Imask.MaskElementm
    masked: masked<T>,
    mask: T['mask'],
    value: string,
    unmaskedValue: string,
    typedValue: Imask.MaskedTypedValue<T['mask']>,
    cursorPos: number,
    readonly selectionStart: number;

    alignCursor(): void;
    alignCursorFriendly(): void;
  }

  export interface IMaskInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
      mask: Imask.AnyMaskedOptions['mask'],
      value: typeof Imask.InputMask['prototype']['value'] | undefined,
      unmask?: boolean;
      radix?: Imask.MaskedNumber['radix']
      overwrite?: typeof Imask.Masked['prototype']['overwrite'],
      placeholderChar?: typeof Imask.MaskedPattern['prototype']['placeholderChar'],
      placeholderChar?: typeof Imask.MaskedPattern['prototype']['lazy'],
      definitions?: typeof Imask.MaskedPattern['prototype']['definitions'],
      blocks?: typeof Imask.MaskedPattern['prototype']['definitions'],
      pattern?: string,
      autofix?: boolean,
      thousandsSeparator?: string,
      mapToRadix?: string[],
      scale?: number,
      signed?: boolean,
      normalizeZeros?: boolean,
      min?: number,
      max?: number,
      onAccept?: (
        value: IMaskInputProps['value'],
        input: InputMask<IMaskInputProps['mask']>
        ) => void;
      onComplete?: (
        value: IMaskInputProps['value'],
        input: InputMask<IMaskInputProps['mask']>,
        e: React.ChangeEvent<HTMLInputElement>
        ) => void;
      inputRef?: (el: HTMLInputElement) => void;
      prepare?: typeof Imask.Masked['prototype']['prepare'],
      validate?: typeof Imask.Masked['prototype']['validate'],
      commit?: typeof Imask.Masked['prototype']['commit'],
      format?: (value: Date) => string,
      parse?: (value: string) => Date,
      dispatch?: typeof Imask.MaskedDynamic['prototype']['dispatch']
  }

  export function IMaskMixin<T, D>(
    Component: React.ComponentType<{ inputRef: React.Ref<D> } & T>
  ): React.ComponentType<T & IMaskInputProps>;

  export class IMaskInput extends React.Component<IMaskInputProps> {}

}
```

The conditional `masked` type is not working properly. When doing `onComplete={(value, input) => {console.log(input.masked)}}`, the type of `masked` is all possible mask types eg: `IMask.MaskedDate | IMask.MaskedNumber | ...` it did not narrow it down to the `mask` prop. If anyone can help, please.

**Comment by vHades0199** [comment]
This work for me.

```ts
declare module 'react-imask' {
  import IMask from 'imask';

  //   /**
  //    * The props accepted by react-imask, based
  //    * on the implementation of imask, with  some additions
  //    */
  export type IMaskInputProps = {
    value?: number | string;
    unmask?: 'typed' | boolean;

    // events
    onAccept?: <T>(value: string, maskRef: IMask.InputMask<IMask.AnyMaskedOptions>, ...args: T[]) => void;
    onComplete?: <T>(value: string, maskRef: IMask.InputMask<IMask.AnyMaskedOptions>, ...args: T[]) => void;

    dispatch?: (
      value: string,
      masked: IMask.MaskedDynamic,
      flags: IMask.AppendFlags
    ) => IMask.AnyMasked | null | undefined;
  } & IMask.AnyMaskedOptions;

  //   /**
  //    * A function that decorates a react component
  //    * with 'imask' props
  //    * @param Component Any React Component
  //    */
  export function IMaskMixin<T, D>(
    Component: React.ComponentType<{ inputRef?: React.Ref<D> } & T>
  ): React.ComponentType<T & IMaskInputProps>;

  //   /**
  //    * A basic IMask React Input
  //    */
  export const IMaskInput: React.ComponentType<IMaskInputProps>;
}

```

**Comment by swashata** [comment]
I also needed to have props being passed to the input. This works for me


```ts
import IMask from 'imask';
import * as React from 'react';

export type AcceptedMaskType = IMask.AnyMaskedOptions['mask'];

export type IMaskProps = {
	mask: AcceptedMaskType;
	value: string;
	onAccept: (value: string, mask: AcceptedMaskType) => void;
	radix: string;
	/**
	 * Placeholder chararcter
	 *
	 * @default '_'
	 */
	placeholderChar?: string;
	overwrite?: boolean;
	unmask: true | false | 'typed';
	/**
	 * Access the inner input HTML.
	 */
	inputRef:
		| React.RefObject<HTMLInputElement>
		| ((instance: HTMLInputElement | null) => void)
		| null
		| undefined;
	/**
	 * If false, placeholder is always visible.
	 *
	 * @default true
	 */
	lazy: boolean;
} & Omit<
	React.DetailedHTMLProps<
		React.InputHTMLAttributes<HTMLInputElement>,
		HTMLInputElement
	>,
	'value' | 'onChange'
>;

export const IMaskInput: React.ComponentType<IMaskProps>;

export type IMaskMixinProps = {
	inputRef:
		| React.RefObject<HTMLInputElement>
		| ((instance: HTMLInputElement | null) => void)
		| null
		| undefined;
} & React.DetailedHTMLProps<
	React.InputHTMLAttributes<HTMLInputElement>,
	HTMLInputElement
>;
export function IMaskMixin(
	binder: (props: IMaskMixinProps) => JSX.Element
): React.ComponentType<IMaskProps>;
```

**Comment by PiTiLeZarD** [comment]
I had to use @OmarOmeiri 's solution just updating onAccept and onComplete using @vHades0199 and that works like a charm, thanks guys!

**Comment by uNmAnNeR** [comment]
hey guys! i started working on it here https://github.com/uNmAnNeR/imaskjs/pull/531
so i did configured build pipeline and tried to add types. As expected there lot of issues appeared that needs to be solved. I've also added React hook as well as vue 3 support. It will be quite valuable release for plugins. The last thing that needs to be done is typings. I do not have much time for it, sorry. So everyone is welcomed to help.

**Comment by uNmAnNeR** [comment]
will be released under v6.2.0
